File: src/main/java/jk_5/nailed/gradle/common/BasePlugin.java
package jk_5.nailed.gradle.common;

import com.google.common.collect.Maps;
import jk_5.nailed.gradle.delayed.DelayedFile;
import jk_5.nailed.gradle.delayed.DelayedFileTree;
import jk_5.nailed.gradle.delayed.DelayedString;
import lombok.Getter;
import org.gradle.api.Action;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.artifacts.repositories.IvyArtifactRepository;
import org.gradle.api.artifacts.repositories.MavenArtifactRepository;

import java.util.Map;

/**
 * No description given
 *
 * @author jk-5
 */
public abstract class BasePlugin implements Plugin<Project> {

    @Getter
    private Project project;

    @Override
    public final void apply(Project project) {
        this.project = project;

        this.project.allprojects(new Action<Project>() {
            @Override
            public void execute(Project project){
                addMavenRepo(project, "reening", "http://maven.reening.nl");
                addMavenRepo(project, "forge", "http://files.minecraftforge.net/maven");
                project.getRepositories().mavenCentral();
                addMavenRepo(project, "minecraft", "https://libraries.minecraft.net");
                addIvyRepo(project, "forgeLegacy", "http://files.minecraftforge.net/[module]/[module]-dev-[revision].[ext]");
            }
        });

        project.afterEvaluate(new Action<Project>() {
            @Override
            public void execute(Project project) {
                BasePlugin.this.afterEvaluate();
            }
        });

        this.applyPlugin();
    }

    public void afterEvaluate(){}
    public abstract void applyPlugin();

    public static void addMavenRepo(Project project, final String name, final String url){
        project.getRepositories().maven(new Action<MavenArtifactRepository>() {
            @Override
            public void execute(MavenArtifactRepository repo){
                repo.setName(name);
                repo.setUrl(url);
            }
        });
    }

    public static void addIvyRepo(Project project, final String name, final String pattern){
        project.getRepositories().ivy(new Action<IvyArtifactRepository>() {
            @Override
            public void execute(IvyArtifactRepository repo){
                repo.setName(name);
                repo.artifactPattern(pattern);
            }
        });
    }

    @SuppressWarnings("unchecked")
    public <T extends Task> T makeTask(String name, Class<T> type){
        Map<String, Object> map = Maps.newHashMap();
        map.put("name", name);
        map.put("type", type);
        return (T) this.getProject().task(map, name);
    }

    protected DelayedString delayedString(String path){
        return new DelayedString(project, path);
    }

    protected DelayedFile delayedFile(String path){
        return new DelayedFile(project, path);
    }

    protected DelayedFileTree delayedFileTree(String path){
        return new DelayedFileTree(project, path);
    }

    protected DelayedFileTree delayedZipTree(String path){
        return new DelayedFileTree(project, path, true);
    }
}

File: src/main/scala/jk_5/nailed/gradle/common/BasePlugin.scala
package jk_5.nailed.gradle.common

import org.gradle.api.{Action, Plugin, Project}
import jk_5.nailed.gradle.delayed.{DelayedFile, DelayedString}
import org.gradle.api.artifacts.repositories.{IvyArtifactRepository, MavenArtifactRepository}
import java.util

/**
 * No description given
 *
 * @author jk-5
 */
trait BasePlugin extends Plugin[Project] {

  private var project: Project = null

  final def apply(project: Project) {
    this.project = project
    this.project.allprojects(new Action[Project] {
      def execute(project: Project) {
        addMavenRepo(project, "reening", "http://maven.reening.nl")
        addMavenRepo(project, "forge", "http://files.minecraftforge.net/maven")
        project.getRepositories.mavenCentral
        addMavenRepo(project, "minecraft", "https://libraries.minecraft.net")
        addIvyRepo(project, "forgeLegacy", "http://files.minecraftforge.net/[module]/[module]-dev-[revision].[ext]")
      }
    })
    project.afterEvaluate(new Action[Project] {
      def execute(project: Project) = BasePlugin.this.afterEvaluate()
    })
    this.applyPlugin()
  }

  def afterEvaluate()
  def applyPlugin()

  def addMavenRepo(project: Project, name: String, url: String) {
    project.getRepositories.maven(new Action[MavenArtifactRepository] {
      def execute(repo: MavenArtifactRepository) {
        repo.setName(name)
        repo.setUrl(url)
      }
    })
  }

  def addIvyRepo(project: Project, name: String, pattern: String) {
    project.getRepositories.ivy(new Action[IvyArtifactRepository] {
      def execute(repo: IvyArtifactRepository) {
        repo.setName(name)
        repo.artifactPattern(pattern)
      }
    })
  }

  @SuppressWarnings(Array("unchecked")) def makeTask[T](name: String, `type`: Class[T]): T = {
    val map: util.Map[String, AnyRef] = new util.HashMap[String, AnyRef]
    map.put("name", name)
    map.put("type", `type`)
    this.getProject.task(map, name).asInstanceOf[T]
  }

  @inline implicit def stringToDelayedString(input: String): DelayedString = this.delayedString(input)
  @inline implicit def stringToDelayedFile(input: String): DelayedFile = this.delayedFile(input)
  @inline protected def delayedString(path: String) = new DelayedString(project, path)
  @inline protected def delayedFile(path: String) = new DelayedFile(project, path)
  @inline protected def getProject = this.project
}
